numpy/core/include/numpy/ufuncobject.h

```diff
@@ -178,8 +190,7 @@ typedef struct _tagPyUFuncObject {
         char *core_signature;
 
         /*
-         * A function which resolves the types and fills an array
-         * with the dtypes for the inputs and outputs.
+         * Required to be NULL (used to be a TypeResolver).
          */
         PyUFunc_TypeResolutionFunc *type_resolver;
```

Comment:
Instead of making this null, could you just have a `UseNoOpsTypeResolver` function you set this to, that does nothing other than call `ufunc->noops_type_resolver(...)` on its arguments?

Reply:
Well, bit of the parsing (for the dtype) creates a new tuple by that time. And creating that tuple seems a bit back and forth (plus takes some time).

Reply:
Arguable it would be better of course, since then downstream could in principle "fall through", but I doubt that such use case exist (and at least it would crash hard).

I am also thinking about starting to merge the loop lookup and type resolver. For some ufuncs there is just duplicate work done there.

Reply:
OK, so writing it down also for myself: After thinking about it a bit more seriously, there is a problem with that approach for user dtypes. Right now, if I have a user dtype it will be valid, but `rational` cannot cast "int7"->rational (since it, as of now) doesn't know about its existence. Of course this is all rather theoretical, but it does show that it is maybe not be a good approach.